R/influxdb.R

```diff
@@ -6,52 +6,52 @@ NULL
 #' 
 #' @param host Character vector with IP address or hostname
 #' @param port Port number
-#' @param username InfluxDB user
-#' @param password InfluxDB password (will be passed in plain text)
 #' @param database The name of the database
 #' @param query Character vector containing the InfluxDB query
-#' @param time_precision Specifies whether the time should be returned in 
-#'   seconds (\code{s}), milliseconds (\code{m}), or microseconds (\code{u}) 
-#'   from epoch (January 1, 1970, 00:00:00 UTC).
 #' @return A named list of data frames, where the names are the series names,
 #'   and the data frames contain the points.
 #'
 #' @export
-influxdb_query <- function(host, port, username, password, database, query,
-                        time_precision=c("s", "m", "u")) {
-  response <- GET(
-    "", scheme = "http", hostname = host, port = port,
-    path = sprintf("db/%s/series", URLencode(database)),
-    query = list(
-      u = username,
-      p = password,
-      q = query,
-      time_precision = match.arg(time_precision),
-      chunked = "false"
-    )
-  )
+influxdb_query <- function (host, port, database, query) {
+  
+  url <- paste0(host, ":", port, "/query?")
+  args <- list(db = database,
+               q = query)
+  
+  getCommand <- httr::GET(url, query=args)
   
   # Check for error. Not familiar enough with httr, there may be other ways it
   # communicates failure.
-  if (response$status_code < 200 || response$status_code >= 300) {
-    if (length(response$content) > 0)
-      warning(rawToChar(response$content))
-    stop("Influx query failed with HTTP status code ", response$status_code)
-  }
+  if (getCommand$status < 200 || getCommand$status >= 300) {
+    
+    stop("Influx query failed with HTTP status code ", getCommand$status)
   
-  response_data <- fromJSON(rawToChar(response$content))
+  } else {
+    
+    response <- httr::content(getCommand, "parsed")
+    if (is.null(response$results[[1]]$error)) {
+      response <- response$results[[1]]$series[[1]]
+      colNames <- unlist(response$columns)
+      nr <- length(response$values)
+      
+      # build the dataframe
+      for (i in 1:nr) {
+        r <- t(unlist(response$values[[i]]))
+        if (i == 1) {
+          output <- data.frame(r, stringsAsFactors = FALSE)
+        } else {
+          output <- rbind(output, data.frame(r))
```

Comment:
(Did the influxdb response format drastically change?)

This is likely an extremely slow way to build up a dataframe (assuming row count >> col count) as every rbind call makes a copy of all the data so far. You'd have to experiment to find the fastest incantation but it usually involves mapping over an entire list using lapply/vapply (as the output vector will then be allocated just once), and then construct the data frame in one step.